app/services/gps_report.rb

```diff
@@ -0,0 +1,131 @@
+# frozen_string_literal: true
+
+class InvalidSupplierError < StandardError; end
+
+class GPSReport
+  include ActionView::Helpers::TextHelper
+
+  def initialize(date_range, supplier_name)
+    @supplier = Supplier.find_by(key: supplier_name)
+    raise InvalidSupplierError, "Invalid supplier: `#{supplier_name}`. Valid values are: `#{Supplier.all.map(&:key).join('`, `')}`." if @supplier.nil?
+
+    @from_date = date_range.first
+    @to_date = date_range.last
+  end
+
+  def generate
+    Rails.logger.info "Generating GPS Data Report for #{@supplier.key} from #{@from_date} to #{@to_date}."
+
+    if journeys.empty?
+      Rails.logger.info "No #{@supplier.key} journeys found for the specified period."
+      return {
+        failures: [],
+        move_count: 0,
+      }
+    end
+
+    Rails.logger.info "Found #{pluralize(moves.count, "#{@supplier.key} move")} with #{pluralize(journeys.count, 'journey')}."
+
+    {
+      failures: failures,
+      move_count: moves.count,
+    }
+  end
+
+private
+
+  def failures
+    return @failures unless @failures.nil?
+
+    @failures = moves.filter_map { |move|
+      { move: move, reason: 'no_journeys' } if move.journeys.empty?
+    }.concat(journey_failures)
```

Comment:
```suggestion
    @failures ||= begin
      moves.filter_map { |move|
        { move: move, reason: 'no_journeys' } if move.journeys.empty?
      }.concat(journey_failures)
    end
```

I think you can do this to avoid needing to check if `@failures` is `nil`, and same for `@moves`.

Reply:
You're right, this one doesn't even need the block :D